src/bin/ion/commands/beta/generate/generator.rs

```diff
@@ -0,0 +1,439 @@
+                    if code_gen_context.data_model == Some(DataModel::Sequence) {
+                        statements.push(format!(
+                            "\"{}\" => {{ data_model.{} =",
+                            &tera_field.name, &tera_field.name,
+                        ));
+                        statements.push(
+                            r#"{
+                let mut values = vec![];
+                reader.step_in()?;
+                while reader.next()? != StreamItem::Nothing {"#
+                                .to_string(),
+                        );
+                        let sequence_type = &tera_field.value.replace("Vec<", "").replace('>', "");
+                        if !self.is_built_in_type(sequence_type) {
+                            statements.push(format!(
+                                "values.push({}::read_from(reader)?)",
+                                sequence_type
+                            ));
+                        } else {
+                            statements.push(format!(
+                                "values.push(reader.read_{}()?)",
+                                sequence_type.to_lowercase()
+                            ));
+                        }
+
+                        statements.push(
+                            r#"}
+                values }}"#
+                                .to_string(),
+                        );
+                    } else if code_gen_context.data_model == Some(DataModel::Value) {
+                        context.insert(
+                            "read_statement",
+                            &format!("{}::read_from(reader)?", &tera_field.value,),
+                        );
+                    } else {
+                        statements.push(format!(
+                            "\"{}\" => {{ data_model.{} = {}::read_from(reader)?;}}",
+                            &tera_field.name, &tera_field.name, &tera_field.value,
+                        ));
+                    }
+                } else {
+                    if code_gen_context.data_model == Some(DataModel::Value) {
+                        context.insert(
+                            "read_statement",
+                            &format!("reader.read_{}()?", &tera_field.value.to_lowercase(),),
+                        );
+                    }
+                    statements.push(format!(
+                        "\"{}\" => {{ data_model.{} = reader.read_{}()?;}}",
+                        &tera_field.name,
+                        &tera_field.name,
+                        &tera_field.value.to_lowercase()
+                    ));
+                }
+            }
+        }
+        context.insert("statements", &statements);
+        Ok(())
+    }
+
+    /// Generates write API for a data model
+    /// This adds statements for writing data model as Ion value that will be used by data model templates
+    // TODO: add support for Java
+    fn generate_write_api(
```

Comment:
The generated write API is going to be so much simpler to implement once @zslayton's new writer API (https://github.com/amazon-ion/ion-rust/pull/680) is released. Maybe we should punt on it until then?

Reply:
+1 to punting, the new code should be much simpler.

Reply:
https://github.com/amazon-ion/ion-cli/issues/72